docs/design-documents/2020-03-logql-fields.md

```diff
@@ -0,0 +1,461 @@
+See below the 3 version for declaring buckets of histogram:
+
+```logql
+{job="app"} | json | histogram latency [.005, .01, .025, .05, .1, .25, .5, 1, 2.5, 5, 10] by (path,status)
+
+{job="app"} | json | histogram size exp(20000, 2, 10) without (instance)
+
+{job="app"} | json | histogram latency linear(.75, 2, 10)
+```
+
+Not providing the array of buckets will result in an error.
+
+>We will only support range vector operations of those series and histograms for now like we do for `count_over_time` and `rate`. This is also because they make more sense as they allow you to select the aggregation range over time for each query. In the future we could may be select the last log line as the instant vector value for each step but I have some doubt it is a useful.
+
+#### Range Vector Operations
+
+Those new functions are very similar to prometheus [over time aggregations](https://prometheus.io/docs/prometheus/latest/querying/functions/#aggregation_over_time). What's interesting with those functions is that they allow the user to pick the range of the aggregation for each steps using the `[1m]` notation.
+
+We plan to support the following new operations:
+
+- `avg_over_time(range-vector)`: the average of all values in the specified interval.
+- `min_over_time(range-vector)`: the minimum of all values in the specified interval.
+- `max_over_time(range-vector)`: the maximum of all values in the specified interval.
+- `sum_over_time(range-vector)`: the sum of all values in the specified interval.
+- `quantile_over_time(scalar, range-vector)`: the φ-quantile (0 ≤ φ ≤ 1) of the values in the specified interval.
+- `stddev_over_time(range-vector)`: the population standard deviation of the values in the specified interval.
+- `stdvar_over_time(range-vector)`: the population standard variance of the values in the specified interval.
+
+> `quantile_over_time(scalar, range-vector)` only works if you have a series with the labels(`le`) this is why you need to use `| histogram` operator on the field stream.
+
+Now let's pretend we have received temperatures from censors around the house into Loki like this:
+
+```log
+2020-03-20T10:20:30 LivingRoom 20C
+2020-03-20T10:21:00 Kitchen 21C
+2020-03-20T10:21:15 Kid1 19C
+....
+```
+
+To get the average temperatures by room in the last 30 minutes we could use the query below:
+
+```logql
+avg_over_time({job="censors"} | glob "* *<room> *<temp>C" | series temp by (room) [30m])
+```
+
+As another example if we have some nginx logs like below:
+
+```log
+127.0.0.1 - - [19/Jun/2012:09:16:22 +0100] "GET /GO.jpg HTTP/1.1" 499 100 "-" "-"
+127.0.0.1 - - [19/Jun/2012:09:16:22 +0100] "GET /test.jpg HTTP/1.1" 200 99 "-" "-"
+```
+
+We could get the 99th percentile latency by path and status over the last 15 minutes using the query below:
+
+```logql
+quantile_over_time(.99, {job="nginx"} | glob "* -- * \"* /*<path> *\" *<status> *<latency> *" | histogram latency linear(.25, 2, 5) by (path, status) [15m])
+```
+
+## API Changes
```

Comment:
@davkal Do you think this is a good approach ? I really wanted to make sure most people even if they don't update can have this new syntax supported.

Reply:
I'd also like some thoughts from 
@davkal on this.